### PR TITLE
Example of an Autolev output test

### PR DESCRIPTION
**Pull request by moorepants**
#### References to other Issues or PRs

Based on #15006.

#### Brief description of what is fixed or changed

This is an example test that ensures the sympy code generated from the autolev parser generates the same output as the autolev code.

@NikhilPappu This is what I'm imagining for output verification tests. Let me know what you think. We should implement some of these for the more complicated examples you have (at least). I'd really like to see this kind of test for all the examples (in addition or instead of the current text comparison tests).

#### Other comments

Note that the position vector test doesn't pass. I think the parser outputs an incorrect result.

#### Release Notes

* parsing
  * Added an example output verification test for the autolev parser.

### CODE REVIEW

File: sympy/parsing/autolev/tests/test_output.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    sympy_input = parse_autolev(autolev_input)
+
+    g = {}
+    l = {}
+    exec(sympy_input, g, l)
+
+    p_o_e = l['point_o'].pos_from(l['point_e']).express(l['frame_d'])
+    v_e_f = l['point_e'].vel(l['frame_f']).express(l['frame_d'])
+    a_e_f = l['point_e'].acc(l['frame_f']).express(l['frame_d'])
```
Comment (moorepants) [inline sympy/parsing/autolev/tests/test_output.py]: The parser set these values but didn't call express. Shouldn't it call express and reassign the values?

**Comment by NikhilPappu** [comment]
@moorepants Did you have a look at the output comparisons of the examples in the [GitLab](https://gitlab.com/sympy/autolev-test-examples) repo?

**Comment by moorepants** [comment]
Here is the code the parser produced:

```python
import sympy.physics.mechanics as me
import sympy as sm
import math as m
import numpy as np

frame_c=me.ReferenceFrame('c')
frame_d=me.ReferenceFrame('d')
frame_f=me.ReferenceFrame('f')
fd, dc = me.dynamicsymbols('fd dc')
fdd, dcd = me.dynamicsymbols('fd dc', 1)
fdd2, dcd2 = me.dynamicsymbols('fd dc', 2)
r, l=sm.symbols('r l', real=True)
point_o=me.Point('o')
point_e=me.Point('e')
frame_d.orient(frame_f, 'Axis', [fd, frame_f.x])
frame_c.orient(frame_d, 'Axis', [dc, frame_d.y])
point_e.set_pos(point_o, r*frame_d.y-l*frame_c.x)
point_e.set_pos(point_o, (point_e.pos_from(point_o)).express(frame_d))
point_e.set_vel(frame_f, ((point_e.pos_from(point_o)).dt(frame_f)).express(frame_d))
point_e.set_acc(frame_f, ((point_e.vel(frame_f)).dt(frame_f)).express(frame_d))
```

I think the correct code should be:

```python
# Note that Autolev users could name variables "sm" and "me", so these could get clobbered.
import sympy as sm
import sympy.physics.mechanics as me
frame_c = me.ReferenceFrame('c')
frame_d = me.ReferenceFrame('d')
frame_f = me.ReferenceFrame('f')
fd, dc = me.dynamicsymbols('fd, dc')
# Use _d for dot and _dd for double dot, etc.
fd_d, dc_d = me.dynamicsymbols('fd, dc', 1)
fd_dd, dc_dd = me.dynamicsymbols('fd, dc', 2)
r, l = sm.symbols('r l', real=True)
point_o = me.Point('o')
point_e = me.Point('e')
# Need to check whether the correct DCMs are generated. Is this F wrt to D or D wrt F? I always forget.
frame_d.orient(frame_f, 'Axis', (fd, frame_f.x))
frame_c.orient(frame_d, 'Axis', (dc, frame_d.y))
point_o.set_pos(point_e, r*frame_d.y - l*frame_c.x)
point_o.set_pos(point_e, point_o.pos_from(point_e).express(frame_d))
point_e.set_vel(frame_f, point_e.pos_from(point_o).dt(frame_f).express(frame_d))
point_e.set_acc(frame_f, point_e.vel(frame_f).df(frame_f).express(frame_d))
```

**Comment by moorepants** [comment]
> @moorepants Did you have a look at the output comparisons of the examples in the GitLab repo?

No, I don't recall being told about them.

File: sympy/parsing/autolev/tests/test_output.py (hunk 1 above)
Comment (moorepants) [reply sympy/parsing/autolev/tests/test_output.py]: Nevermind, this is done in the parser.

**Comment by NikhilPappu** [comment]
@moorepants I had added output comparisons to some of the examples in the GitLab repo. They aren't very detailed though. In many cases, I have just shown the equivalence (I have only included comments in the example files) of the final eoms.
Here is one of them (from the chaos_pendulum example):
```
# Output Comparison Results
# =========================

# -> (51) ZERO[1] = 0
# -> (52) ZERO[2] = 0
# -> (53) ZERO[3] = -G*LB*MB*SIN(THETA) - G*MA*LA*SIN(THETA) - 2*(IBXX-IBYY)*SIN(PHI)*COS(PHI)*ALPHA*OMEGA - (MB*LB^2+IAYY+IBYY+MA*LA^2+(IBXX-IBYY)*SIN(PHI)^2)*OMEGA'
# -> (54) ZERO[4] = (IBXX-IBYY)*SIN(PHI)*COS(PHI)*OMEGA^2 - IBZZ*ALPHA'

subs_dict = {g:9.81,lb:0.2,w:0.2,h:0.1,ma:0.01,mb:0.1,theta:np.deg2rad(90),phi:np.deg2rad(0.5),omega:1,alpha:2,omegad:3,alphad:4}
zero_autolev = sm.Matrix([[0],[0]])
zero_autolev[0] = (-g*lb*mb*sm.sin(theta) - g*ma*la*sm.sin(theta) - 2*(ibxx-ibyy)*sm.sin(phi)*sm.cos(phi)*alpha*omega -\
(mb*lb**2+iayy+ibyy+ma*la**2+(ibxx-ibyy)*sm.sin(phi)**2)*omegad).subs(subs_dict).evalf()

zero_autolev[1] = ((ibxx-ibyy)*sm.sin(phi)*sm.cos(phi)*omega**2-ibzz*alphad).subs(subs_dict).evalf()
print(zero_autolev)
# Matrix([[-0.217020788909953], [-0.00133624206773955]])

zero[0] = zero[0].subs(subs_dict).evalf()
zero[1] = zero[1].subs(subs_dict).evalf()
print(zero)
# Matrix([[-0.217020788909953], [-0.00133624206773955]])
# which shows that zero_autolev and zero are the same equations.
```

**Comment by moorepants** [comment]
I am not able to find output examples in the Gitlab repo. Can you point me specifically to one (or them)?

**Comment by NikhilPappu** [comment]
@moorepants I had just put the comparisons (mainly comments) at the end of the example files.

**Comment by moorepants** [comment]
I see. Ideally we encode them as I've done in this PR, as also ideally have as much as we can in the SymPy repo. Note that on my first one I seemingly uncovered an error(s) in the parser output.

**Comment by NikhilPappu** [comment]
@moorepants 
These tests certainly look better. The main reason I have included them in the GitLab repo is that the outputs were related to the examples (most of which are from Autolev texts). I think we could keep them there are make GitLab-CI run those tests by using SymPy.
As for the error, I shall fix it soon as the fix is easy.

**Comment by moorepants** [comment]
Note that the P_O_E> parsed output is still incorrect.

**Comment by NikhilPappu** [comment]
@moorepants Now that I took another look at the position vectors, I think the parser does it the right way.
`p_o_e>` is the position vector of `e` wrt `o`.
```
>>> import sympy.physics.mechanics as me
>>> import sympy as sm
>>> import math as m
>>> import numpy as np

>>> n = me.ReferenceFrame('n')
>>> o = me.Point('p')
>>> e = me.Point('o')
>>> l = sm.symbols('l')
# p_o_e> = l*n.x
>>> e.set_pos(o, l*n.x)
>> e.pos_from(o)
l*n.x
```

**Comment by moorepants** [comment]
You are correct. I forgot the syntax. Here is a snippet from Online Dynamics that confirms:
selection_263
The position vector goes from the first second character to the third character.

Sorry to lead you astray.

**Comment by moorepants** [comment]
In the doctest builds I'm getting:

```
$ bin/test_travis.sh
+[[ true == \t\r\u\e ]]
+python bin/test_setup.py
Traceback (most recent call last):
  File "bin/test_setup.py", line 27, in <module>
    assert setup.tests == test_list, set(setup.tests).symmetric_difference(set(test_list))
AssertionError: {'sympy.parsing.autolev.tests'}
```

Not quite sure what that is.

**Comment by NikhilPappu** [comment]
@moorepants I placed the current test file `test_autolev.py` in `parsing/tests` and put the comparison files in `parsing/autolev/test_examples` so there is no directory called `parsing/autolev/tests`.
So `parsing/autolev/tests` needs to be added in the `modules = [` and `tests = [` lists of `setup.py`.

File: sympy/parsing/autolev/tests/test_output.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+P_O_E>=EXPRESS(P_O_E>,D)
+V_E_F>=EXPRESS(DT(P_O_E>,F),D)
+A_E_F>=EXPRESS(DT(V_E_F>,F),D)\
+"""
+
+    sympy_input = parse_autolev(autolev_input)
+
+    g = {}
+    l = {}
+    exec(sympy_input, g, l)
```
[the remark this answers is not in the document]
Comment (NikhilPappu) [reply sympy/parsing/autolev/tests/test_output.py]: Not sure of another way either.
@asmeurer

**Comment by moorepants** [comment]
This is ready for final review and merge.

**Comment by moorepants** [comment]
@NikhilPappu Would you mind reviewing this?

**Comment by NikhilPappu** [comment]
@moorepants It looks great to me!

**Comment by NikhilPappu** [comment]
@ayushmankoul 
```
self.output_code.append("import sympy.physics.mechanics as me\n")
self.output_code.append("import sympy as sm\n")
self.output_code.append("import math as m\n")
self.output_code.append("import numpy as np\n")
```
These should also have `_me`, `_sm_ and `_np`. Make this change, do the `test_autolev.py` thing to update the tests and you should be good to go for the most part.

**Comment by ayushmankoul** [comment]
@NikhilPappu I updated this thing in code but I am not able to run skipped tests on my local machine.The following code in `test_autolev.py` is causing assertion errors:

assertionerror
Whereas on altering the above code as following yields no assertion error on local machine:

noerror

Sorry for messing things up again and again.

**Comment by NikhilPappu** [comment]
The changed one is just to update the existing test files. It will
obviously pass as it isn't testing anything. The changes should be reverted
after running the second one. Can you post the Assertion errors here?

**Comment by moorepants** [comment]
It'd be best to move this discussion to the respective issue or PR.

**Pull request merged**